pkg/operator/mon/mon.go

```diff
@@ -464,16 +401,83 @@ func (c *Cluster) loadMonConfig() error {
 	return nil
 }
 
-// detect whether we have a big enough cluster to run services on different nodes.
-// the anti-affinity will prevent pods of the same type of running on the same node.
-func (c *Cluster) getAntiAffinity() (bool, error) {
+// detect the nodes that are available for new mons to start.
+func (c *Cluster) getAvailableMonNodes() ([]v1.Node, error) {
 	nodeOptions := metav1.ListOptions{}
 	nodeOptions.TypeMeta.Kind = "Node"
 	nodes, err := c.context.Clientset.CoreV1().Nodes().List(nodeOptions)
 	if err != nil {
-		return false, fmt.Errorf("failed to get nodes in cluster. %+v", err)
+		return nil, err
+	}
+	logger.Infof("there are %d nodes available for mons (existing mons=%d)", len(nodes.Items), len(c.clusterInfo.Monitors))
+
+	// get the nodes that have mons assigned
+	nodesInUse, err := c.getNodesWithMons()
+	if err != nil {
+		logger.Warningf("could not get nodes with mons. %+v", err)
+		nodesInUse = map[string]interface{}{}
+	}
+
+	// choose nodes for the new mons that don't have mons currently
+	availableNodes := []v1.Node{}
+	for _, node := range nodes.Items {
+		if _, ok := nodesInUse[node.Name]; !ok && validNode(node) {
+			availableNodes = append(availableNodes, node)
+		}
+	}
+	logger.Infof("Found %d running nodes without mons", len(availableNodes))
+
+	// if all nodes already have mons, just add all nodes to be available
+	if len(availableNodes) == 0 {
+		logger.Infof("All nodes are running mons. Adding all %d nodes to the availability.", len(nodes.Items))
+		for _, node := range nodes.Items {
+			if validNode(node) {
+				availableNodes = append(availableNodes, node)
+			}
+		}
+	}
+	if len(availableNodes) == 0 {
+		return nil, fmt.Errorf("no nodes are available for mons")
+	}
+
+	return availableNodes, nil
+}
+
+func validNode(node v1.Node) bool {
+	// a node must be ready to be valid for mon placement
+	for _, c := range node.Status.Conditions {
+		if c.Type == v1.NodeReady {
+			return true
+		}
+	}
+	logger.Infof("node %s is not ready. %+v", node.Name, node.Status.Conditions)
+	return false
+}
+
+func (c *Cluster) getNodesWithMons() (map[string]interface{}, error) {
```

Comment:
are you using `map[string]interface{}` here as a Set?  could we use `util.Set` instead if that's the case?